### PR TITLE
CLDC-1433 Add parent child relationship model

### PR DESCRIPTION
**Pull request by natdeanlewissoftwire**
Tested locally
Risk of another migration from same base at the same time - should be checked before merge
No docs required
Ticket: https://digital.dclg.gov.uk/jira/browse/CLDC-1443

### CODE REVIEW

File: db/migrate/20220810152340_add_parent_child_relationships.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class AddParentChildRelationships < ActiveRecord::Migration[7.0]
+  def change
+    create_table :organisation_relationships do |t|
+      t.integer :child_organisation_id, foreign_key: true
+      t.integer :parent_organisation_id, foreign_key: true
+      t.timestamps
```
Comment (natdeanlewissoftwire) [inline db/migrate/20220810152340_add_parent_child_relationships.rb:6]: Apparently this is good practice, but current tables in the database seem inconsistent as to whether they have timestamp data - do we want this?

File: db/schema.rb [hunk 2]
```diff
@@ -383,6 +390,7 @@
   add_foreign_key "case_logs", "organisations", column: "owning_organisation_id", on_delete: :cascade
   add_foreign_key "case_logs", "schemes"
   add_foreign_key "locations", "schemes"
+  add_foreign_key "schemes", "organisations", column: "managing_organisation_id"
```
Comment (natdeanlewissoftwire) [inline db/schema.rb]: Already merged in, not sure why it's showing up here as this is now conistent with main

File: app/models/organisation.rb [hunk 3]
```diff
@@ -6,6 +6,10 @@ class Organisation < ApplicationRecord
   has_many :organisation_rent_periods
   has_many :owned_schemes, class_name: "Scheme", foreign_key: "owning_organisation_id", dependent: :delete_all
   has_many :managed_schemes, class_name: "Scheme", foreign_key: "managing_organisation_id"
+  has_many :parent_organisation_relationships, foreign_key: :child_organisation_id, class_name: "OrganisationRelationship"
```
Comment (jamesrose) [inline app/models/organisation.rb:9]: @natdeanlewissoftwire Could you add some model tests for this? I think at a minimum we would like to test whether an organisation can have parents and children.

File: db/migrate/20220810152340_add_parent_child_relationships.rb (hunk 1 above)
Comment (jamesrose) [reply db/migrate/20220810152340_add_parent_child_relationships.rb:6]: I generally keep timestamps on these kinds of join tables. It adds very little overhead, but it's useful to debug when a relationship was created.

File: db/schema.rb (hunk 2 above)
Comment (jamesrose) [reply db/schema.rb]: Do you just need to rebase your branch?

File: db/migrate/20220810152340_add_parent_child_relationships.rb (hunk 1 above)
Comment (baarkerlounger) [reply db/migrate/20220810152340_add_parent_child_relationships.rb:6]: I think all existing tables have timestamps? Which ones don't?

File: db/schema.rb (hunk 2 above)
Comment (natdeanlewissoftwire) [reply db/schema.rb]: Of course, thanks :)

File: db/migrate/20220810152340_add_parent_child_relationships.rb (hunk 1 above)
Comment (natdeanlewissoftwire) [reply db/migrate/20220810152340_add_parent_child_relationships.rb:6]: @baarkerlounger Ah you're right, they all have created_at at least - it's the couple that don't have updated_at columns I was thinking of (logs_exports and versions), never mind!

Comment (baarkerlounger) [reply db/migrate/20220810152340_add_parent_child_relationships.rb:6]: `versions` isn't really ours, it comes from https://github.com/paper-trail-gem/paper_trail

File: spec/factories/organisation.rb [hunk 4]
```diff
@@ -11,17 +11,15 @@
     holds_own_stock { true }
   end
 
-  factory :organisation_la do
-    organisation
-    ons_code { "E07000041" }
-    created_at { Time.zone.now }
-    updated_at { Time.zone.now }
-  end
-
```
Comment (natdeanlewissoftwire) [inline spec/factories/organisation.rb:20]: this is the (I _think_) unused factory I was referring to @jamesrose

File: spec/factories/organisation.rb [hunk 5]
```diff
@@ -11,17 +11,15 @@
...
     organisation
     rent_period { 2 }
     created_at { Time.zone.now }
     updated_at { Time.zone.now }
   end
+
+  factory :organisation_relationship do
+    child_organisation { FactoryBot.create(:organisation) }
+    parent_organisation { FactoryBot.create(:organisation) }
+  end
```
Comment (natdeanlewissoftwire) [inline spec/factories/organisation.rb:24]: I put this back in the organisation file, I think that's simpler than having a separate organisation_relationship file especially as organisation_rent_period is here too

File: spec/models/organisation_spec.rb [hunk 6]
```diff
@@ -27,6 +27,22 @@
         .to raise_error(ActiveRecord::RecordInvalid, "Validation failed: Provider type #{I18n.t('validations.organisation.provider_type_missing')}")
     end
 
+    context "with parent/child association" do
+      let(:child_organisation) { FactoryBot.create(:organisation, name: "DLUHC_Child") }
```
Comment (jamesrose) [inline spec/models/organisation_spec.rb]:
```suggestion
      let(:child_organisation) { FactoryBot.create(:organisation, name: "DLUHC Child") }
```

nit: underscored name

**Review by jamesrose — APPROVED** [review]
